### ISSUE TITLE
CUDA: Performance issues with `torch.unique`

### ISSUE THREAD

**Issue by Marsmaennchen221** [pytorch/pytorch#122274]
### 🐛 Describe the bug

I've noticed some performance problems with `torch.unique`. I tested the following:
```py
import timeit

import torch

number = 10

t = torch.randint(10, (1_000_000, ), device="cpu")
t_cuda = t.to("cuda")

print(timeit.timeit("t.unique(dim=0)", globals=globals(), number=number))
print(timeit.timeit("t_cuda.unique(dim=0)", globals=globals(), number=number))

t = torch.randint(10, (1_000, 1_000), device="cpu")
t_cuda = t.to("cuda")

print(timeit.timeit("t.unique(dim=0)", globals=globals(), number=number))
print(timeit.timeit("t_cuda.unique(dim=0)", globals=globals(), number=number))

t = torch.randint(10, (10, 100, 1_000), device="cpu")
t_cuda = t.to("cuda")

print(timeit.timeit("t.unique(dim=0)", globals=globals(), number=number))
print(timeit.timeit("t_cuda.unique(dim=0)", globals=globals(), number=number))
```
Output:
```py
27.928257999941707
0.323013799963519
0.03842799994163215
0.05541190004441887
0.004202899988740683
5.427228400018066
```

It makes sense that the tests get better since it compares a million values in the first test, a thousand in the second, and only ten different tensors in the third. The unexpected behavior is that the cuda implementation somehow gets worse on the last test.

### Versions

PyTorch version: 2.2.0+cu121                                                                                     
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro
GCC version: (x86_64-posix-seh-rev2, Built by MinGW-W64 project) 12.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.1 (tags/v3.11.1:a7a450f, Dec  6 2022, 19:58:39) [MSC v.1934 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: 12.1.66
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080
Nvidia driver version: 546.65
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=3701
DeviceID=CPU0
Family=107
L2CacheSize=6144
L2CacheSpeed=
Manufacturer=AuthenticAMD
MaxClockSpeed=3701
Name=AMD Ryzen 9 5900X 12-Core Processor
ProcessorType=3
Revision=8450

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] torch==2.2.0+cu121
[pip3] torchvision==0.17.0+cu121
[conda] Could not collect

**Comment by tringwald** [comment]
I can reproduce this on `2.2` and `main`. Seems like CUDA has a problem with >2-dim tensors when `unique` is calculated over a dimension other than the last one.

**Comment by malfet** [comment]
By the way, using `timeit` is not the right way to benchmark anything on GPU, as one will benchmark either just time required to schedule the operator on GPU, or memcpy time. Right way to do the benchmarking is using specialized tools that just record time events between time kernel was scheduled until the time its execution finished, see https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html

**Comment by msaroufim** [comment]
FWIW I could repro a dramatic slowdown for the 3D case `10x100x1000` more specifically the slowdown is egregious for small first dimensions but not totally sure if the issue is launch overhead or insufficient parallelism

```
- dim0=10: CPU is 144.78x faster 
- dim0=100: CPU is 4.12x faster 
- dim0=1000: CUDA is 7.91x faster 
- dim0=10000: CUDA is 61.92x faster 
```
